File: 2903-insert-greatest-common-divisors-in-linked-list/2903-insert-greatest-common-divisors-in-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
   public int gcd(int a, int b) {
        if (b == 0) 
            return a;
        return gcd(b, a % b);
    }

    
    public ListNode insertGreatestCommonDivisors(ListNode head) {
        if (head == null || head.next == null)  // Check if the list is empty or has only one node
            return head;
        
        ListNode current = head;
        ListNode current1 = head.next;

        while (current1 != null) {
            int gc = gcd(current.val, current1.val);  // Calculate GCD
            ListNode temp = new ListNode(gc);  // Create a new node with GCD value
            temp.next = current1;  // Insert the new node between current and current1
            current.next = temp;
            
            current = current1;  // Move to the next node
            current1 = current1.next;  // Move to the next node
        }
        
        return head;
        
    }
}

File: 2903-insert-greatest-common-divisors-in-linked-list/2903-insert-greatest-common-divisors-in-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int gcd(int a, int b) {

        if (b == 0)
            return a;
        return gcd(b, a % b);
    }
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        ListNode* current = head;
        if (head == NULL || head->next == NULL)
            return head;
        ListNode* current1 = head->next;
        while (current1 != NULL) {
            int gc = gcd(current->val, current1->val);
            ListNode* temp = new ListNode(gc);
            temp->next = current1;
            current->next = temp;
            current = current1;
            current1 = current1->next;
        }
        return head;
    }
};